### ISSUE TITLE
Refresh wildcard content in debug/exploded mode.

### ISSUE THREAD

**Issue by simoami** [Orange-OpenSource/wro4j-taglib#10]
In development mode we set exploded to true. We use wildcards to specify the content of various groups. If we introduce a new file or delete an existing one, there's no way for our app to pick it up, say on browser refresh. A server restarted. While this is expected in production environments, it's painful to have to restart the server every time.

In debug mode, It would be good to re-read the file system for any changes wherever wildcards are used.

**Comment by tatablack** [comment]
I agree about the need to be able to refresh the configuration without restarting the server. I'm not so sure whether we can consider `exploded=true` as equivalent to a "debug mode", though for the majority of cases this will probably hold true.

Anyway, I see the following options:
- use `exploded` or a new configuration parameter to trigger configuration reload _at every request_ of a static resource. In theory, users of the tag library should not have a lot of resources in a page (that would defeat wro4j's purpose), but I think it might be a little overkill.
- add a new configuration parameter (`reload=true|false`), maybe with an additional, optional, parameter to set the delay between configuration reloads. These would happen not on every request but at fixed intervals. However, this should happen in a separate thread, which would need to be cleaned up when needed (i.e. when the web app gets reloaded).
- expose reloading functionality through JMX.
- add `reload=true` to the url, in order to trigger a configuration reload when the first wro4j tag is encountered (not particularly elegant, but should be simple to implement).

Comments, preferences, alternatives, suggestions...?

**Comment by alexo** [comment]
There is a configuration property available since 1.4.8, called resourceWatcherUpdatePeriod which helps to detect the changes in development mode. This property helps to avoid unnecessary processing if there is no change.

Since you are using wildcards in your model, probably you should also invalidate the model periodically to ensure that the exploded wildcards contains latest resources. Therefore, you could try using modelUpdatePeriod=5 which will ensure that each 5 seconds you will have a fresh model. Setting resourceWatcherUpdatePeriod=10 will ensure that each 10 seconds you'll get the most recent content.

Read more about [Configurations Options](http://code.google.com/p/wro4j/wiki/ConfigurationOptions)

**Comment by tatablack** [comment]
Alex, thanks, I completely forgot about `resourceWatcherUpdatePeriod`. It would be nice to leverage it, instead of trying to rebuild similar functionality. I can think of a couple of issues, though:
- wro4j-taglib, as far as I can see, loads wro4j model only at initialization time (using [ServletContextAttributeHelper](https://github.com/Orange-OpenSource/wro4j-taglib/blob/wro4j-taglib-1.3/src/main/java/com/orange/wro/taglib/config/WroConfig.java#L74)). This of course needs to be changed, somehow;
- would the "resource watcher" work if the web app does not use wro4j's filter, but only WroServletContextListener?

**Comment by alexo** [comment]
> wro4j-taglib, as far as I can see, loads wro4j model only at initialization time (using ServletContextAttributeHelper). This of course needs to be changed, somehow;

The WroManager uses WroModelFactory to load the model each time the processing is performed. As long as the model is built in WroModelFactory and is not mutated anywhere else (I'm planning to make the WroModel immutable in future releases, to avoid unexpected behavior when custom code mutates the model), everything should work fine.

> would the "resource watcher" work if the web app does not use wro4j's filter, but only WroServletContextListener?

The resource watcher is a feature which is always available. It doesn't matter if you use filter or not. So the answer is yes, it should work even when using WroServletContextListener.
